extensions-contrib/gce-extensions/src/main/java/org/apache/druid/indexing/overlord/autoscaling/gce/GceAutoScaler.java

```diff
@@ -0,0 +1,539 @@
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This module permits the autoscaling of the workers in GCE
+ *
+ * General notes:
+ * - The IPs are IPs as in Internet Protocol, and they look like 1.2.3.4
+ * - The IDs are the names of the instances of instances created, they look like prefix-abcd,
+ *   where the prefix is chosen by you and abcd is a suffix assigned by GCE
+ */
+@JsonTypeName("gce")
+public class GceAutoScaler implements AutoScaler<GceEnvironmentConfig>
+{
+  private static final EmittingLogger log = new EmittingLogger(GceAutoScaler.class);
+
+  private final GceEnvironmentConfig envConfig;
+  private final int minNumWorkers;
+  private final int maxNumWorkers;
+  private final SimpleWorkerProvisioningConfig config;  // For future use
+
+  private Compute cachedComputeService = null;
+
+  private static final long POLL_INTERVAL_MS = 5 * 1000;  // 5 sec
+  private static final int RUNNING_INSTANCES_MAX_RETRIES = 10;
+  private static final int OPERATION_END_MAX_RETRIES = 10;
+
+  @JsonCreator
+  public GceAutoScaler(
+          @JsonProperty("minNumWorkers") int minNumWorkers,
+          @JsonProperty("maxNumWorkers") int maxNumWorkers,
+          @JsonProperty("envConfig") GceEnvironmentConfig envConfig,
+          @JacksonInject SimpleWorkerProvisioningConfig config
+  )
+  {
+    Preconditions.checkArgument(minNumWorkers > 0,
+                                "minNumWorkers must be greater than 0");
+    this.minNumWorkers = minNumWorkers;
+    Preconditions.checkArgument(maxNumWorkers > 0,
+                                "maxNumWorkers must be greater than 0");
+    Preconditions.checkArgument(maxNumWorkers > minNumWorkers,
+                                "maxNumWorkers must be greater than minNumWorkers");
+    this.maxNumWorkers = maxNumWorkers;
+    this.envConfig = envConfig;
+    this.config = config;
+  }
+
+  /**
+   * CAVEAT: this is meant to be used only for testing passing a mock version of Compute
+   */
+  @VisibleForTesting
+  public GceAutoScaler(
+          int minNumWorkers,
+          int maxNumWorkers,
+          GceEnvironmentConfig envConfig,
+          SimpleWorkerProvisioningConfig config,
+          Compute compute
+  )
+  {
+    this(minNumWorkers, maxNumWorkers, envConfig, config);
+    this.cachedComputeService = compute;
+  }
+
+  @Override
+  @JsonProperty
+  public int getMinNumWorkers()
+  {
+    return minNumWorkers;
+  }
+
+  @Override
+  @JsonProperty
+  public int getMaxNumWorkers()
+  {
+    return maxNumWorkers;
+  }
+
+  @Override
+  @JsonProperty
+  public GceEnvironmentConfig getEnvConfig()
+  {
+    return envConfig;
+  }
+
+  private synchronized Compute createComputeService()
+      throws IOException, GeneralSecurityException, InterruptedException, GceServiceException
+  {
+    final int maxRetries = 5;
+
+    int retries = 0;
+    while (cachedComputeService == null && retries < maxRetries) {
```

Comment:
The retries are on `Compute.Builder.build()` returning `null` and not on the `Compute.Builder.build()` throwing exceptions.  In the second case we just log and propagate the exception.  It has been noticed that, at least at start up, the `Compute.Builder.build()` may return `null`.

Unfortunately the documentations of the GCE API is not prolix enough to explain why / how this can happen (from here to grand-parent):

- https://developers.google.com/resources/api-libraries/documentation/compute/v1/java/latest/com/google/api/services/compute/Compute.Builder.html#build--
- https://googleapis.dev/java/google-api-client/latest/com/google/api/client/googleapis/services/json/AbstractGoogleJsonClient.Builder.html#build--
- https://googleapis.dev/java/google-api-client/latest/com/google/api/client/googleapis/services/AbstractGoogleClient.Builder.html#build--